pkg/processor/build/runtime/java/runtime.go

```diff
@@ -0,0 +1,364 @@
+	return fmt.Sprintf("nuclio/handler-java:%s-%s",
+		j.versionInfo.Label,
+		j.versionInfo.Arch), nil
+}
+
+// "reverser.jar:Reverser" -> "Reverser"
+func (j *java) handlerClassName(handler string) string {
+	fields := strings.Split(handler, ":")
+	if len(fields) == 1 {
+		return fields[0]
+	}
+	return fields[1]
+}
+
+func (j *java) createBuildFile(stagingBuildDir string) error {
+	buildFilePath := path.Join(stagingBuildDir, buildFileName)
+	if common.IsFile(buildFilePath) {
+		j.Logger.InfoWith("Found user build file, using it", "path", buildFilePath)
+		return nil
+	}
+
+	buildTemplate, err := template.New("build").Parse(buildTemplateCode)
+	if err != nil {
+		return err
+	}
+
+	buildFile, err := os.Create(buildFilePath)
+	if err != nil {
+		return err
+	}
+
+	defer buildFile.Close()
+
+	data := map[string]interface{}{
+		"Dependencies": j.FunctionConfig.Spec.Build.Dependencies,
+		"Handler":      j.handlerClassName(j.FunctionConfig.Spec.Handler),
+	}
+	return buildTemplate.Execute(buildFile, data)
+}
+
+func (j *java) createDockerFile(dockerfilePath, imageName string) error {
+	if !j.FunctionConfig.Spec.Build.NoBaseImagesPull {
+		// pull the onbuild image we need to build the processor builder
+		if err := j.DockerClient.PullImage(imageName); err != nil {
+			return errors.Wrap(err, "Failed to pull onbuild image for java")
+		}
+	}
+
+	dockerFileContent := fmt.Sprintf("FROM %s", imageName)
+
+	return ioutil.WriteFile(dockerfilePath, []byte(dockerFileContent), 0600)
+}
+
+func (j *java) isFile(filePath, extension string) bool {
+	return common.IsFile(filePath) && strings.ToLower(path.Ext(filePath)) == strings.ToLower(extension)
+}
+
+func (j *java) createJavaSourceDir() error {
+	// Java sources *must* be under src/main/java
+	// TODO: Should we use gradle's sourceSets in current directory?
+	// https://docs.gradle.org/current/userguide/java_plugin.html
+	// (Probably no since there might be more than one Java file in the root directory)
+	javaSrcDirPath := path.Join(j.StagingDir, "src/main/java")
+
+	if err := os.MkdirAll(javaSrcDirPath, 0777); err != nil {
+		return errors.Wrap(err, "Can't create Java source directory")
+	}
+
+	buildPath := j.FunctionConfig.Spec.Build.Path
+	var filesToCopy []string
+	var err error
+	if common.IsFile(buildPath) {
+		filesToCopy = append(filesToCopy, buildPath)
+	} else {
+		filesToCopy, err = filepath.Glob(path.Join(buildPath, "*.java"))
+		if err != nil {
+			return errors.Wrapf(err, "Can't find Java files in %q", buildPath)
+		}
+	}
+
+	if len(filesToCopy) == 0 {
+		return errors.Errorf("Can't find Java files in %q", buildPath)
+	}
+
+	for _, filePath := range filesToCopy {
+		destPath := path.Join(javaSrcDirPath, path.Base(filePath))
+		if err := util.CopyFile(filePath, destPath); err != nil {
+			return errors.Wrap(err, "Can't copy file to Java source directory")
+		}
+	}
+
+	return nil
+}
+
+func (j *java) copyJavaSources(buildPath, stagingBuildDir string) error {
```

Comment:
I don't understand what the outcome of this function is - please comment heavily

Reply:
Done